File: 707_software_patterns/assn/01/src/swift/Patterns/Patterns/XMLSerializer.swift
//
//  XMLSerializer.swift
//  Patterns
//
//  Created by Jon Lindberg on 12/31/18.
//  Copyright © 2018 Jon Lindberg. All rights reserved.
//

import Foundation

public class XMLSerializer
{
	init?(filename : String)
	{
		//
		// There must be a way to create the file and get a handle to writing to it in one call without
		// jumping down to direct 'C' Posix calls, but I haven't found it yet.
		//
		if !FileManager.default.createFile(atPath: filename, contents: nil, attributes: nil)
		{
			print("Unable to create file '\(filename)'.")
			return nil
		}
		
		do
		{
			let tempFile = try FileHandle(forWritingTo: URL(fileURLWithPath: filename))
			file	= tempFile
		}
		catch
		{
			print(error)
			return nil
		}
	}
	
	internal var indentationLevel	: Int	= 0
	internal var file		: FileHandle
	
	func prettyIndentation()
	{
		if let data = "\t".data(using: String.Encoding.utf8)
		{
			for _ in 0..<indentationLevel
			{
				file.write(data)
			}
		}
	}
	public func serializePretty(node : Node)
	{
		if let doc = node as? Document
		{
			if let header = """
			  <? xml version="1.0" encoding="UTF-8"?>
			  """.data(using: String.Encoding.utf8)
			{
				file.write(header)
				
				if let documentElement = doc.documentElement
				{
					serializePretty(node: documentElement)
				}
			}
		}
		else if let elem = node as? Element
		{
			prettyIndentation()
			if let openTag = ("<" + elem.tagName).data(using: String.Encoding.utf8)
			{
				file.write(openTag)
			}
			
			var attrCount	= 0
			
			for attr in elem.attributes
			{
				serializePretty(node: attr)
				attrCount += 1
			}
			
			if attrCount > 0
			{
				if let space = " ".data(using: String.Encoding.utf8)
				{
					file.write(space)
				}
			}
			
			if elem.childNodes.count == 0
			{
				if let closer = "/>\n".data(using: String.Encoding.utf8)
				{
					file.write(closer)
				}
			}
			else
			{
				if let closer = ">\n".data(using: String.Encoding.utf8)
				{
					file.write(closer)
				}
				
				indentationLevel += 1
					
				for child in elem.childNodes
				{
					if (child as? Element) != nil || (child as? Text) != nil
					{
						serializePretty(node: child)
					}
				}

				indentationLevel -= 1
				prettyIndentation()
				
				if let endTag = ("</" + elem.tagName + ">\n").data(using: String.Encoding.utf8)
				{
					file.write(endTag)
				}
			}
		}
		else if let attr = node as? Attr
		{
			if let attrValue = (" " + attr.name + "=\"" + attr.value + "\"").data(using: String.Encoding.utf8)
			{
				file.write(attrValue)
			}
		}
		else if let text = node as? Text
		{
			prettyIndentation()
			if let data = (text.data + "\n").data(using: String.Encoding.utf8)
			{
				file.write(data)
			}
		}
	}
	public func serializeMinimal(node : Node)
	{
		if let doc = node as? Document
		{
			if let header = "<? xml version=\"1.0\" encoding=\"UTF-8\"?>".data(using: String.Encoding.utf8)
			{
				file.write(header)
				
				if let documentElement = doc.documentElement
				{
					serializeMinimal(node: documentElement)
				}
			}
		}
		else if let elem = node as? Element
		{
			if let openTag = ("<" + elem.tagName).data(using: String.Encoding.utf8)
			{
				file.write(openTag)
			}
			
			for attr in elem.attributes
			{
				serializeMinimal(node: attr)
			}
			
			if elem.childNodes.count == 0
			{
				if let closer = "/>".data(using: String.Encoding.utf8)
				{
					file.write(closer)
				}
			}
			else
			{
				if let closer = ">".data(using: String.Encoding.utf8)
				{
					file.write(closer)
				}
				
				for child in elem.childNodes
				{
					if (child as? Element) != nil || (child as? Text) != nil
					{
						serializeMinimal(node: child)
					}
				}
				
				if let endTag = ("</" + elem.tagName + ">").data(using: String.Encoding.utf8)
				{
					file.write(endTag)
				}
			}
		}
		else if let attr = node as? Attr
		{
			if let attrValue = (" " + attr.name + "=\"" + attr.value + "\"").data(using: String.Encoding.utf8)
			{
				file.write(attrValue)
			}
		}
		else if let text = node as? Text
		{
			if let data = text.data.data(using: String.Encoding.utf8)
			{
				file.write(data)
			}
		}
	}
}

File: 707_software_patterns/assn/02/prog/src/XMLSerializer.cpp
#include "XMLSerializer.hpp"

#include "Attr.hpp"
#include "Document.hpp"
#include "Element.hpp"
#include "Text.hpp"

void XMLSerializerStrategy::prettyIndentation() {

    for (int i = 0; i < indentationLevel; i++) {
        file << "\t";
    }
}

void XMLSerializerPretty::serialize(dom::Node* node) {

    if (dynamic_cast<dom::Document*>(node) != 0) {

        file << "<? xml version=\"1.0\" encoding=\"UTF-8\"?>";
        file << "\n";
        serialize(dynamic_cast<dom::Document*>(node)->getDocumentElement());

    } else if (dynamic_cast<dom::Element*>(node) != 0) {

        prettyIndentation();
        file << "<" << dynamic_cast<dom::Element*>(node)->getTagName();

        int attrCount = 0;

        for (dom::NamedNodeMap::iterator i =
                 dynamic_cast<dom::Element*>(node)->getAttributes()->begin();
             i != dynamic_cast<dom::Element*>(node)->getAttributes()->end();
             i++) {
            serialize(*i);
            attrCount++;
        }

        if (attrCount > 0)
            file << " ";

        if (dynamic_cast<dom::Element*>(node)->getChildNodes()->size() == 0) {
            file << "/>";
            file << "\n";
        } else {

            file << ">";
            file << "\n";
            indentationLevel++;

            for (dom::NodeList::iterator i = dynamic_cast<dom::Element*>(node)
                                                 ->getChildNodes()
                                                 ->begin();
                 i != dynamic_cast<dom::Element*>(node)->getChildNodes()->end();
                 i++) {
                if (dynamic_cast<dom::Element*>(*i) != 0 ||
                    dynamic_cast<dom::Text*>(*i) != 0) {
                    serialize(*i);
                }
            }
            indentationLevel--;
            prettyIndentation();
            file << "</"
                 << dynamic_cast<dom::Element*>(node)->getTagName() + ">";
            file << "\n";
        }
    } else if (dynamic_cast<dom::Attr*>(node) != 0) {
        file << " " << dynamic_cast<dom::Attr*>(node)->getName() << "=\""
             << dynamic_cast<dom::Attr*>(node)->getValue() << "\"";
    } else if (dynamic_cast<dom::Text*>(node) != 0) {
        prettyIndentation();
        file << dynamic_cast<dom::Text*>(node)->getData();
        file << "\n";
    }
}

void XMLSerializerMinimal::serialize(dom::Node* node) {

    if (dynamic_cast<dom::Document*>(node) != 0) {

        file << "<? xml version=\"1.0\" encoding=\"UTF-8\"?>";
        serialize(dynamic_cast<dom::Document*>(node)->getDocumentElement());

    } else if (dynamic_cast<dom::Element*>(node) != 0) {

        file << "<" << dynamic_cast<dom::Element*>(node)->getTagName();

        for (dom::NamedNodeMap::iterator i =
                 dynamic_cast<dom::Element*>(node)->getAttributes()->begin();
             i != dynamic_cast<dom::Element*>(node)->getAttributes()->end();
             i++) {
            serialize(*i);
        }
        if (dynamic_cast<dom::Element*>(node)->getChildNodes()->size() == 0) {

            file << "/>";

        } else {

            file << ">";

            for (dom::NodeList::iterator i = dynamic_cast<dom::Element*>(node)
                                                 ->getChildNodes()
                                                 ->begin();
                 i != dynamic_cast<dom::Element*>(node)->getChildNodes()->end();
                 i++)
                if (dynamic_cast<dom::Element*>(*i) != 0 ||
                    dynamic_cast<dom::Text*>(*i) != 0)
                    serialize(*i);

            file << "</"
                 << dynamic_cast<dom::Element*>(node)->getTagName() + ">";
        }

    } else if (dynamic_cast<dom::Attr*>(node) != 0) {

        file << " " << dynamic_cast<dom::Attr*>(node)->getName() << "=\""
             << dynamic_cast<dom::Attr*>(node)->getValue() << "\"";

    } else if (dynamic_cast<dom::Text*>(node) != 0) {
        file << dynamic_cast<dom::Text*>(node)->getData();
    }
}

// void XMLSerializer::serialize(bool minimal, dom::Node* node) {

//     if (dynamic_cast<dom::Document*>(node) != 0) {
//         file << "<? xml version=\"1.0\" encoding=\"UTF-8\"?>";
//         if (!minimal) {
//             file << "\n";
//         }
//         serialize(minimal,
//                   dynamic_cast<dom::Document*>(node)->getDocumentElement());
//     } else if (dynamic_cast<dom::Element*>(node) != 0) {
//         if (!minimal) {
//             prettyIndentation();
//         }
//         file << "<" << dynamic_cast<dom::Element*>(node)->getTagName();

//         int attrCount = 0;

//         for (dom::NamedNodeMap::iterator i =
//                  dynamic_cast<dom::Element*>(node)->getAttributes()->begin();
//              i != dynamic_cast<dom::Element*>(node)->getAttributes()->end();
//              i++) {
//             serialize(minimal, *i);
//             if (!minimal) {
//                 attrCount++;
//             }
//         }

//         if (!minimal) {
//             if (attrCount > 0) {
//                 file << " ";
//             }
//         }

//         if (dynamic_cast<dom::Element*>(node)->getChildNodes()->size() == 0)
//         {
//             file << "/>";
//             if (!minimal) {
//                 file << "\n";
//             }
//         } else {
//             file << ">";
//             if (!minimal) {
//                 file << "\n";
//                 indentationLevel++;
//             }

//             for (dom::NodeList::iterator i =
//             dynamic_cast<dom::Element*>(node)
//                                                  ->getChildNodes()
//                                                  ->begin();
//                  i !=
//                  dynamic_cast<dom::Element*>(node)->getChildNodes()->end();
//                  i++) {
//                 if (dynamic_cast<dom::Element*>(*i) != 0 ||
//                     dynamic_cast<dom::Text*>(*i) != 0) {
//                     serialize(minimal, *i);
//                 }
//             }
//             if (!minimal) {
//                 indentationLevel--;
//                 prettyIndentation();
//             }
//             file << "</"
//                  << dynamic_cast<dom::Element*>(node)->getTagName() + ">";
//             if (!minimal) {
//                 file << "\n";
//             }
//         }
//     } else if (dynamic_cast<dom::Attr*>(node) != 0) {
//         file << " " << dynamic_cast<dom::Attr*>(node)->getName() << "=\""
//              << dynamic_cast<dom::Attr*>(node)->getValue() << "\"";
//     } else if (dynamic_cast<dom::Text*>(node) != 0) {
//         if (!minimal) {
//             prettyIndentation();
//         }
//         file << dynamic_cast<dom::Text*>(node)->getData();
//         if (!minimal) {
//             file << "\n";
//         }
//     }
// }
